src/app/icd/server/ICDManager.cpp

```diff
@@ -234,15 +240,87 @@ bool ICDManager::CheckInMessagesWouldBeSent()
             }
 
             // At least one registration would require a Check-In message
-            VerifyOrReturnValue(mSubInfoProvider->SubjectHasActiveSubscription(entry.fabricIndex, entry.monitoredSubject), true);
+            VerifyOrReturnValue(!shouldCheckInMsgsBeSentFunction(entry.fabricIndex, entry.monitoredSubject), true);
         }
     }
 
     // None of the registrations would require a Check-In message
     return false;
 }
 
-void ICDManager::TriggerCheckInMessages()
+/**
+ * ShouldCheckInMsgsBeSentAtActiveModeFunction is used to determine if a Check-In message is required for a given registration.
+ * Due to how the ICD Check-In use-case interacts with the persistent subscription and subscription timeout resumption features,
+ * having a single implementation of the function renders the implementation very difficult to understand and maintain.
+ * Because of this, each valid feature combination has its own implementation of the function.
+ */
+#if CHIP_CONFIG_PERSIST_SUBSCRIPTIONS
+#if CHIP_CONFIG_SUBSCRIPTION_TIMEOUT_RESUMPTION
+/**
+ * @brief Implementation for when the persistant subscription and subscription timeout resumption feature are present.
+ *        Function checks that there no active or persisted subscriptions for a given fabricIndex or subjectID.
+ *
+ * @param aFabricIndex
+ * @param subjectID subjectID to check. Can be an operational node id or a CAT
+ *
+ * @return true Returns true if the fabricIndex and subjectId combination does not have an active or a persisted subscription.
+ * @return false Returns false if the fabricIndex and subjectId combination has an active or persisted subscription.
+ */
+bool ICDManager::ShouldCheckInMsgsBeSentAtActiveModeFunction(FabricIndex aFabricIndex, NodeId subjectID)
+{
+    return !(mSubInfoProvider->SubjectHasActiveSubscription(aFabricIndex, subjectID) ||
+             mSubInfoProvider->SubjectHasPersistedSubscription(aFabricIndex, subjectID));
+}
+#else
+/**
+ * @brief Implementation for when the persistant subscription feature is present without the subscription timeout resumption
+ * feature. Function checks that there no active subscriptions. If the boot up subscription resumption has not been completed,
+ *        function also checks if there are persisted subscriptions.
+ *
+ * @note The persistent subscriptions feature tries to resume subscriptions at the highest min interval
```

Comment:
Right, this sort of comment is what we need in the case above.

Reply:
done